src/cloudflare/internal/pipeline-transform.ts

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2024 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+import entrypoints from 'cloudflare-internal:workers'
+
+async function* readLines(stream: ReadableStream<string>): AsyncGenerator<string> {
+  let start, end = 0
+  let partial = ''
+
+  // @ts-ignore
+  for await (const chunk of stream) {
+    const full = chunk+partial as string
+    for (const char of full) {
+      if (char === '\n') {
+        yield chunk.substring(start, end)
+        end++
+        start = end
+      } else {
+        end++
+      }
+    }
+
+    partial = chunk.substring(start, end)
+    start = 0
+    end = 0
+  }
+
+  if (partial.length > 0) {
+    yield partial
+  }
+}
+
+type Batch = {
+  id: string 	// unique identifier for the batch
+  shard?: string 	// assigned shard
+  ts: number 	// timestamp of the event
+
+  format: Format
+  data: unknown
+}
+
+type JsonStream = Batch & {
+  format: Format.JSON_STREAM,
+  data: ReadableStream<Uint8Array>
+}
+
+enum Format {
+  JSON_STREAM = 'json_stream', //jsonl
+}
+
+export class PipelineTransformImpl extends entrypoints.WorkerEntrypoint {
+  #batch?: Batch
+  #initalized: boolean = false
+
+  constructor(ctx: unknown, env: unknown) {
+    super(ctx, env)
+  }
+
+  // stub overriden on the sub class
+  // @ts-ignore
+  public async transformJson(data: object[]): Promise<object[]> {
+    throw new Error('should be implemented by parent')
+  }
+
+  // called by the dispatcher which then calls the subclass methods
+  // @ts-ignore
+  private async _transform(batch: Batch): Promise<JsonStream> {
```

Comment:
Since we check for `batch == null`, we should update the type of the variable.

```suggestion
  private async _transform(batch?: Batch): Promise<JsonStream> {
```

Reply:
Went ahead and made this a separate method instead